Port OG-Context module as OgContext provider

pfrenssen opened the pull request:
This is a continuation of https://github.com/amitaibu/og/pull/252. It is a simplified version of the initial implementation of OgContext that was created in that PR.

Commit:
First sketch of the OG context module.

Commit:
Adding files.

Commit:
Moving OG context to OG core.

Commit:
Commit more work.

Commit:
Iterate over plugins easily.

Commit:
Get list of plugins by a logic you desire.

Commit:
Minor fixes.

Commit:
Get the plugins in the right order.

Commit:
Get the group from the url.

Commit:
Pushing work.

Commit:
Get the group from a group content.

Commit:
Get all the groups.

Commit:
Remove a line in the end of the file.

Commit:
My own CR fixes.

Commit:
Adding some tests.

Commit:
Merge master

Commit:
CS fixes(hope)

Commit:
More CS.

Commit:
she's into a typical riff.

Commit:
Hmm...

Commit:
Return the first group for now.

Commit:
Keep entity as the only plugin.

Commit:
Testing the context it self.

Commit:
Rename the plugin to GroupResolver.

Commit:
Refactor the entity name [skip ci]

Commit:
More renaming [skip ci]

Commit:
More renaming [skip ci]

Commit:
Down with renaming.

Commit:
Fix code sniffer.

Commit:
Merge branch '8.x-1.x' into 242

Commit:
Fix tests.

Commit:
Fix conflicts

Commit:
Extend from the correct class and fix a service.

Commit:
Make CS happy.

Commit:
Adding test.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into 242

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into 242

pfrenssen committed:
Simplify the way we will resolve groups to determine group context.

pfrenssen committed:
Use the standard naming scheme for plugin managers.

pfrenssen committed:
Provide an updated interface for group resolver plugins.

pfrenssen committed:
Rename the GroupResolver plugin type to OgGroupResolver.

pfrenssen committed:
Use correct name for the property containing a reference to the group type manager service.

claudiu-cristea commented on src/Plugin/OgGroupResolver/Entity.php:
```diff
@@ -0,0 +1,91 @@
+      }
+
+      if ($this->groupTypeManager->isGroup($parameter->getEntityTypeId(), $parameter->bundle())) {
+        return $parameter;
```
The interface states that ::getGroups() should return EntityInterface[] but here returns a single item. Shouldn't be `return [$parameter];`?

pfrenssen replied:
This is still left from the original approach, this is still adhering to the previous implementation of the interface.

I think this plugin should probably be entirely removed. This is not a reliable way of discovering entities on routes. I am planning on using the entity paths instead.

This code as well as the test code is still to be refactored, it's better to wait a bit with reviewing. I can't manage the labels on this issue, but I'll remove the "WIP" from the title and post a comment  when it is ready to be reviewed.

amitaibu commented on src/OgGroupResolverInterface.php:
```diff
@@ -0,0 +1,67 @@
+   * Use this if the plugin has determined the relevant group in the current
+   * context with 100% certainty.
+   */
+  public function stopPropagation();
```
nice solution - 👍

pfrenssen replied:
This was an idea from @sandervd! There are some cases where you really don't want other plugins influencing the outcome, for example if you are on the canonical path for a group,  or if you have a dedicated plugin written for a project that handles the resolving in a custom way.

amitaibu commented on src/Plugin/OgGroupResolver/Entity.php:
```diff
@@ -0,0 +1,91 @@
+        continue;
+      }
+
+      return Og::getGroups($parameter);
```
I think it's better to inject `MembershipManager` directly (would make unit tests a bit easier)

pfrenssen replied:
I am thinking of replacing this plugin with a more reliable solution. We can get a list of entity paths from the entity definition and check if the current route matches any of these entity paths. This would avoid false positives on routes that have multiple arguments and happen to have a group in one of them.

Imagine you have a search page that allows to filter on related content:

> /search/related/{entity}/{entity}/{entity}

Imagine the 3rd entity is an article that happens to be group content. If we blindly loop over the route parameters then we will suddenly decide that this search results page belongs to the group of the 3rd entity.

This could cause all kinds of unexpected behaviour.

amitaibu commented on tests/modules/og_test/src/Plugin/GroupResolver/UserGroupReference.php:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Get the group from a field reference attached to the current user.
+ *
+ * @GroupResolver(
```
=> `@OgGroupResolver`? (Just because there could be other `group` of different things - happened already in D7 even with the vertical tabs module or something similar)

pfrenssen replied:
:+1:  I had already fixed this but I hadn't yet pushed my changes.

amitaibu reviewed (commented):
Haven't reviewed all, but just a few small comments

pfrenssen committed:
Initial untested implementation of OgContext.

pfrenssen committed:
OgGroupResolver plugins need to declare which cache contexts are affected by their results.

pfrenssen committed:
Update plugin name, this has been renamed to OgGroupResolver.

pfrenssen committed:
Document how the discovery of the group context works.

pfrenssen committed:
Exploring the first OgResolver plugin.

pfrenssen committed:
Fix bugs popping up after actually trying to run the code :)

pfrenssen committed:
Add schema for the group resolvers configuration entry.

pfrenssen committed:
Initial version of the RouteGroupContentResolver.

pfrenssen committed:
Initial version of the RequestQueryArgumentResolver.

pfrenssen committed:
Use the correct method name.

pfrenssen commented:
@amitaibu in https://github.com/amitaibu/og/pull/252#issuecomment-247079122 you outline the algorithm to resolve the groups:

```
OgContextProvider::getRuntimeContexts() will call the OgGroupResolver plugins, and let them return all their values. That is, it will not break on the first group that is found.

Then, it will apply some simple (hardcoded) rule to get the "Best matching" group, by order:

1. User has that group in the session, because in the previous page requests they go it as context + user has access to it. This is for cases when yuu previously visited a group content that belonged to group A, and now you visit a group content that belongs to groups A and B -- so we should be nice and keep you in the A context.
2. User has view access to the group
3. The first rule
```

The "view access to the group" is a bit problematic in combination with plugins returning their own best candidate. It is possible that a plugin has two potential candidates, but the "best" one is not accessible by the user, but the second best candidate is. I think it would be better that the plugins always return the full set of groups, and simply order them so that the best candidate is the first result. This would mean to get rid of `OgGroupResolverInterface::getBestCandidate()`.

amitaibu commented:
> The "view access to the group" is a bit problematic in combination with plugins returning their own best candidate

In OG7, and I think we can keep in OG8 for simplicity the `getBestCandidate` should be done once on the top level by the ContextProvider . That is, the plugins will return an _un-ordered_ list of groups. Then it would be the job of `getBestCandidate` to iterate over them, filter out the non-accessible ones, and choose one single group.

pfrenssen commented:
OK, that sounds like a great approach, I'll rework it to work like this.

pfrenssen commented:
@amitaibu I realized something else: the session approach won't work out of the box for anonymous users on Drupal 8. For performance reasons anonymous users no longer have a cookie, so they don't have a session.

It is still possible to initiate a session for anonymous users, but it would be highly inappropriate for OG to do this out of the box, since it would make any website that has OG enabled to be uncacheable by reverse proxies such as Varnish.

I'm guessing that for this particular PR it is OK not to worry about this. We can come up with a good solution in the future, e.g. rewriting outgoing links to add query arguments, or allow to opt-in to sessions. I just wanted to mention this. I'll also add a mention of it in the documentation of the user session plugin.

pfrenssen committed:
Don't let plugins decide which is the best candidate. This is the responsibility of the resolver.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into 242-simplified

pfrenssen committed:
Remove the tests for the initial approach. They are no longer valid.

pfrenssen committed:
The original Entity plugin has been replaced by RouteGroupResolver and RouteGroupContentResolver.

pfrenssen committed:
Fix coding standards.

pfrenssen committed:
Account for the difference between plugins that provide actual groups, and ones that filter existing groups.

pfrenssen committed:
Start on unit tests for the OgGroupResolver plugins.

pfrenssen committed:
Test RouteGroupResolver::getGroups().

pfrenssen committed:
Fix PHP_CodeSniffer warnings.

pfrenssen committed:
Moar PHP_CodeSniffer warning fixes.

pfrenssen committed:
Remove the notion of a OgResolver plugin type.

pfrenssen committed:
Initial implementation of the OgResolvedGroupCollection.

pfrenssen committed:
Use an OgResolvedGroupCollection to gather the resolved groups, instead of returning them.

pfrenssen committed:
Convert RouteGroupResolver to use OgResolvedGroupCollection.

pfrenssen committed:
Partially start using the OgResolvedGroupCollection in the context provider.

pfrenssen committed:
Fix PHP_CodeSniffer warnings.

pfrenssen commented:
This was discussed IRL between @amitaibu and me at Drupalcon. We are going to get rid of the distinction between two "types" of plugins ("providers" vs "selectors"). Instead we will have a single plugin type that can either provide groups, or remove groups that were discovered by previous plugins. This will work by passing an object to the plugins that contains a collection of the previously discovered groups.

This will simplify the approach.

pfrenssen committed:
Refactor group resolver test so its functionality can be reused in other tests.

pfrenssen committed:
Convert RouteGroupContentResolver to use the group collection.

pfrenssen committed:
Start on a test for the RouteGroupContentResolver.

pfrenssen committed:
Convert RequestQueryArgumentResolver to use the resolved group collection.

pfrenssen committed:
Update documentation.

pfrenssen committed:
Merge remote-tracking branch 'gizra/8.x-1.x' into 242-simplified

pfrenssen commented on tests/src/Unit/Plugin/OgGroupResolver/OgRouteGroupResolverTestBase.php:
```diff
@@ -0,0 +1,317 @@
+      // @todo Provide a mock for a call to ??::isGroupContent() as soon as this
+      //   is available in a service.
+      // $this->groupContentTypeManager??->isGroupContent($entity_type_id, $bundle_id)
+      //  ->willReturn($is_group_content);
```
I'm using unit tests to test the plugins, but I hit a blocker: I need to be able to check if an entity is group content, but this is currently not possible in OO code. It is currently handled by `Og::isGroupContent()` which delegates to a static method `OgGroupAudienceHelper::getAllGroupAudienceFields()`. I cannot mock this :-1:

It seems like `OgGroupAudienceHelper` is one of the final bastions of procedural code left in OG. It only contains 2 methods so it will probably be a very quick conversion.

I will make a separate ticket for this.

pfrenssen committed:
Fix coding standards violation.

pfrenssen commented:
I have converted `OgGroupAudienceHelper` into a service in #199. I will merge that branch in here so this is unblocked.

pfrenssen committed:
Merge branch 'convert-group-audience-helper-to-service' into 242-simplified

pfrenssen committed:
Merge remote-tracking branch 'gizra/8.x-1.x' into 242-simplified

pfrenssen committed:
Finish the unit test of the RouteGroupContentResolver now that GroupAudienceHelper has been converted to OO.

pfrenssen committed:
Store all properties in a single array.

This will make them easier to manage and to return to the consuming code.

pfrenssen committed:
Store cache context information in the resolved group collection.

pfrenssen committed:
Remove empty method.

pfrenssen committed:
Refactor shared code for the route based plugin tests to the base class.

pfrenssen committed:
Do not forget to call your parent.

pfrenssen committed:
Maintain identical ordering for the dependencies.

pfrenssen committed:
Let the base test class handle dependency injection.

pfrenssen committed:
Refactor code to make it reusable in tests.

pfrenssen committed:
Add a test for the RequestQueryArgumentResolver plugin.

pfrenssen commented:
The approach may be simplified but these tests are far from simple :sweat_smile:

pfrenssen commented:
I must say I find it much more rewarding to write unit tests as opposed to kernel tests, just the fact that they run so fast makes it much more interesting to work on. I always lose my focus if my test takes longer than a few seconds to run.

amitaibu commented:
I also like unit tests, as they allow you to perform own code review really well.

amitaibu commented on src/ContextProvider/OgContext.php:
```diff
@@ -0,0 +1,209 @@
+
+    // @todo Move this in the collection? Seems to make sense, since the
+    //   collection has knowledge about the groups and votes.
+    // Find the best matching group by iterating over the candidates and return
```
so, is `votes` like +1 that each plugin may give?

pfrenssen replied:
Yes, with a small difference, their vote will be depending on their position in the list, so the first plugin's vote will count for more than the second plugin's vote. This will allow us to easily choose the correct "winner" if we have two groups with the same number of votes.

pfrenssen replied:
Note that the OgContext code still needs to be cleaned up and tested, it still contains parts from the original implementation. But the basic idea will remain the same.

pfrenssen committed:
Change method name to better reflect what it actually returns.

pfrenssen committed:
Move the hardcoded filtering by user access to a dedicated plugin.

pfrenssen committed:
Refactor the sorting of the resolved groups into the collection class.

pfrenssen committed:
Simplify the handling of cache contexts now that the heavy lifting is done by the collection.

pfrenssen committed:
Update documentation.

pfrenssen committed:
Fix coding standards.

pfrenssen committed:
Rename property so that it is clearer what it contains.

pfrenssen committed:
Make it possible to override the way mocked test group entities are created during tests.

pfrenssen committed:
Mark the method that returns test entity properties as abstract so that it is clear that tests need to implement this.

pfrenssen committed:
Update documentation.

pfrenssen committed:
Add a test for the UserGroupAccessResolver plugin.

pfrenssen committed:
Fix coding standards.

pfrenssen commented:
This still needs a test for `OgResolvedGroupCollection` and for `OgContext` itself. `OgResolvedGroupCollectionTest` will be a very straightforward unit test, but I'm not yet sure what would be the best approach to test `OgContext`.

I feel hesitant to write a functional test for this, since we would have to provide a test module that contains "something" (e.g. a block or a page controller) that consumes the available context and outputs data about it that we can check. For me providing a test module specifically for this seems overkill. Also I have more fun things to do in my spare time than waiting for functional tests to run :grin:

A unit test will be quite complicated since much of the business logic in in protected methods, so this would mean that we need to do a ton of mocking.

Maybe a kernel test would be a solution, but this functionality relies on requests being made, which is normally in the domain of functional tests. I don't know how to 'fake' a request in a kernel test.

I will investigate this further in the weekend. For the moment I'm pretty happy with how this is turning out. It also seems to work quite nice in practice. I just replaced our project's custom context provider with this and ran our entire Behat test suite against it and it came back green.

pfrenssen committed:
Start on a test for OgResolvedGroupCollection.

pfrenssen committed:
Add more test coverage for OgResolvedGroupCollection.

pfrenssen committed:
Test the sorting of groups in a collection and fix bugs discovered by the test.

pfrenssen committed:
Sorting items with the same value is undefined, we can't test this.

This was giving different results on different versions of PHP. The PHP manual
states: "If two members compare as equal, their relative order in the sorted
array is undefined.".

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into 242-simplified

pfrenssen committed:
Start on a test for OgContext.

amitaibu commented:
This will require a re-roll now.

pfrenssen commented:
Thanks for letting me know. Going to pick this back up!

pfrenssen committed:
Merge remote-tracking branch 'upstream/8.x-1.x' into 242-simplified

pfrenssen committed:
Start testing the logic in OgContext::getBestCandidate().

pfrenssen committed:
Add a standard test case, a single group resolver returning a single group.

pfrenssen commented:
Haha thanks for the compliment! :) I'm having great fun writing these, it's like a bunch of little puzzles to solve, and I'm learning so much about the internal workings of things like typed data.

This test is not very clean, but this is due to `ContextProviders` not using dependency injection. They use typed data which they get from `\Drupal::typedDataManager()` so I have to put a mocked `TypedDataManager` on the container.

I'm currently sprinting at a local open source conference so there is a good chance I will be able to finally finish this today. Fingers crossed :)

pfrenssen committed:
Fix PHP_CodeSniffer warnings.

pfrenssen committed:
Test if cache contexts are correctly provided in the cacheability metadata.

pfrenssen committed:
Add a test case for two resolver plugins that each return a different result.

pfrenssen committed:
Add a complex test case, and test if propagation can be stopped.

pfrenssen commented:
Alright that went pretty smooth! I think this is ready!

@amitaibu care for a look?

amitaibu commented:
Yay! (Hectic days so might take a couple of days for me to have time to do _proper_ review.)

amitaibu commented:
I saw two `todo`s there, but there's really no reason to not get this in. Thank you!!
